## Final Project Submission

Please fill out:
* Student name: Jeremy Nguyo 
* Student pace: full time
* Scheduled project review date/time: 
* Instructor name: Nikita Njoroge
* Blog post URL:


##  DATA PREPARATION

## Importing libraries we may need for our analysis 

In [1]:
# pandas for data analysis and manipulation
import pandas as pd
# numpy for numerical operations
import numpy as np
# seaborn and matplotlib for data visualization
import seaborn as sns
import matplotlib.pyplot as plt
# sqlite library for connect to sql database
import sqlite3

## Importing the dataframes

In [2]:
# connect to the sqlite 'im' database file from IMDb
conn = sqlite3.connect('zippedData/im.db')
# view the list of tables in the database
tables_db = pd.read_sql("""SELECT name FROM sqlite_master WHERE type = 'table';""", conn)
tables_db

name
0   movie_basics
1      directors
2      known_for
3     movie_akas
4  movie_ratings
5        persons
6     principals
7        writers

#### For our analysis we will use two tables for the IMDb database. Namely: movie_basics and movie_ratings

In [3]:
#Preview of all the columns and the first five rows of the movie_basics table
movie_basics_df = pd.read_sql(""" SELECT * FROM movie_basics; """,conn)
movie_basics_df.head()

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [4]:
#Checking to see the datatype, number of columns and rows of the movie_basics table
movie_basics_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


In [5]:
#understanding the parameters we'll be working with
movie_basics_df.nunique()

movie_id           146144
primary_title      136071
original_title     137773
start_year             19
runtime_minutes       367
genres               1085
dtype: int64

In [6]:
#checking for null values in the movie_basic_df
movie_basics_df.isnull().sum()

movie_id               0
primary_title          0
original_title        21
start_year             0
runtime_minutes    31739
genres              5408
dtype: int64

In [7]:
#Preview of all the columns and the first five rows of the movie_ratings table
movie_ratings_df = pd.read_sql(""" SELECT * FROM movie_ratings; """,conn)
movie_ratings_df.head()

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [8]:
#Checking to see the datatype, number of columns and rows of the movie_ratings table
movie_ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   movie_id       73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


In [9]:
#understanding the parameters we'll be working with
movie_ratings_df.nunique()

movie_id         73856
averagerating       91
numvotes          7349
dtype: int64

In [10]:
#checking for null values in the movie_ratings_df
movie_ratings_df.isnull().sum()

movie_id         0
averagerating    0
numvotes         0
dtype: int64

In [11]:
#understanding our data
movie_ratings_df.describe()

averagerating      numvotes
count   73856.000000  7.385600e+04
mean        6.332729  3.523662e+03
std         1.474978  3.029402e+04
min         1.000000  5.000000e+00
25%         5.500000  1.400000e+01
50%         6.500000  4.900000e+01
75%         7.400000  2.820000e+02
max        10.000000  1.841066e+06

#### We will also use the dataset from tn movie budgets for our analysis

In [12]:
#Preview of all the columns and the first five rows of the bom dataset
movie_budgets_df = pd.read_csv('zippedData/tn.movie_budgets.csv')
movie_budgets_df.head()

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  
0      $425,000,000   $760,507,625  $2,776,345,279  
1      $410,600,000   $241,063,875  $1,045,663,875  
2      $350,000,000    $42,762,350    $149,762,350  
3      $330,600,000   $459,005,868  $1,403,013,963  
4      $317,000,000   $620,181,382  $1,316,721,747

In [13]:
#Checking to see the datatype, number of columns and rows of the movie_budgets table
movie_budgets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


In [14]:
#understanding the parameters we'll be working with
movie_budgets_df.nunique()

id                    100
release_date         2418
movie                5698
production_budget     509
domestic_gross       5164
worldwide_gross      5356
dtype: int64

In [15]:
#checking for null values in the movies_budgets_df
movie_budgets_df.isnull().sum()

id                   0
release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
dtype: int64

In [16]:
#understanding our data
movie_budgets_df.describe()

id
count  5782.000000
mean     50.372363
std      28.821076
min       1.000000
25%      25.000000
50%      50.000000
75%      75.000000
max     100.000000

## DATA CLEANING

Checking for Validity, Accuracy, Completeness,Consistency and Uniformity of the Data

In the movie_basics_df, we will need the genres column and it seems to have some missing values 

In [17]:
#Checking for null values
movie_basics_df['genres'].isnull().sum()

5408

In [18]:
#lets have a look at these missing values
#Filter to select only the missing values
filt = movie_basics_df['genres'].isnull()

#view the missing values and assign them to a variable
genres_missing_values_df =  movie_basics_df.loc[filt]
genres_missing_values_df

movie_id                                      primary_title  \
16      tt0187902  How Huang Fei-hong Rescued the Orphan from the...   
22      tt0253093                                       Gangavataran   
35      tt0306058                                      Second Coming   
40      tt0326592                                      The Overnight   
44      tt0330811                                Regret Not Speaking   
...           ...                                                ...   
146088  tt9907396                Footloose in the Cotswolds - Part 1   
146089  tt9907608                Footloose in the Cotswolds - Part 2   
146107  tt9910922       Doctor Who Augmented Reality: Times Magazine   
146129  tt9914942                         La vida sense la Sara Amat   
146142  tt9916730                                             6 Gunn   

                                           original_title  start_year  \
16      How Huang Fei-hong Rescued the Orphan from the...        2011   
22                                           Gangavataran        2018   
35                                          Second Coming        2012   
40                                          The Overnight        2010   
44                                    Regret Not Speaking        2011   
...                                                   ...         ...   
146088                Footloose in the Cotswolds - Part 1        2016   
146089                Footloose in the Cotswolds - Part 2        2016   
146107       Doctor Who Augmented Reality: Times Magazine        2013   
146129                         La vida sense la Sara Amat        2019   
146142                                             6 Gunn        2017   

        runtime_minutes genres  
16                  NaN   None  
22                134.0   None  
35                 95.0   None  
40                 88.0   None  
44                  NaN   None  
...                 ...    ...  
146088            118.0   None  
146089            102.0   None  
146107              NaN   None  
146129              NaN   None  
146142            116.0   None  

[5408 rows x 6 columns]

In [19]:
#Determining what percentage of rows in this column contain missing values
total_num_rows = len(movie_basics_df['genres'])
total_num_missing_rows = len(genres_missing_values_df['genres'])
percentage = total_num_missing_rows/total_num_rows
percentage

0.037004598204510616

The genre is quite a unique value. You can not just randomly assign a genre to a movie. Being that there are only a few missing values, 3.7% of the total, then i think dropping these rows would be more reasonable

In [20]:
movie_basics_df.dropna(subset='genres')

movie_id                                primary_title  \
0       tt0063540                                    Sunghursh   
1       tt0066787              One Day Before the Rainy Season   
2       tt0069049                   The Other Side of the Wind   
3       tt0069204                              Sabse Bada Sukh   
4       tt0100275                     The Wandering Soap Opera   
...           ...                                          ...   
146138  tt9916428                          The Secret of China   
146139  tt9916538                          Kuambil Lagi Hatiku   
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
0                                         Sunghursh        2013   
1                                   Ashad Ka Ek Din        2019   
2                        The Other Side of the Wind        2018   
3                                   Sabse Bada Sukh        2018   
4                             La Telenovela Errante        2017   
...                                             ...         ...   
146138                          The Secret of China        2019   
146139                          Kuambil Lagi Hatiku        2019   
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes                 genres  
0                 175.0     Action,Crime,Drama  
1                 114.0        Biography,Drama  
2                 122.0                  Drama  
3                   NaN           Comedy,Drama  
4                  80.0   Comedy,Drama,Fantasy  
...                 ...                    ...  
146138              NaN  Adventure,History,War  
146139            123.0                  Drama  
146140              NaN            Documentary  
146141              NaN                 Comedy  
146143              NaN            Documentary  

[140736 rows x 6 columns]

In [21]:
#dropping the rows with the missing values in the genres column
movie_basics_df.dropna(subset='genres', inplace=True)
#checking to see if the rows have been dropped
movie_basics_df['genres'].isnull().sum()

0

In the movie_budgets_df, the Id column has repetitive values. See below

In [22]:
print('Number of unique values in the id column are:', movie_budgets_df['id'].nunique(), 'out of a total of:', len(movie_budgets_df['id']))

Number of unique values in the id column are: 100 out of a total of: 5782


Since this column has no other use other than being a unique identifier,It is wise to drop it since it doesn't serve it's purpose. We can use the names column as the index once we inspect it further since it seems to have duplicates.

In [23]:
#dropping the id column
movie_budgets_df.drop(columns='id', inplace=True)
movie_budgets_df

release_date                                        movie  \
0     Dec 18, 2009                                       Avatar   
1     May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2      Jun 7, 2019                                 Dark Phoenix   
3      May 1, 2015                      Avengers: Age of Ultron   
4     Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   
...            ...                                          ...   
5777  Dec 31, 2018                                       Red 11   
5778   Apr 2, 1999                                    Following   
5779  Jul 13, 2005                Return to the Land of Wonders   
5780  Sep 29, 2015                         A Plague So Pleasant   
5781   Aug 5, 2005                            My Date With Drew   

     production_budget domestic_gross worldwide_gross  
0         $425,000,000   $760,507,625  $2,776,345,279  
1         $410,600,000   $241,063,875  $1,045,663,875  
2         $350,000,000    $42,762,350    $149,762,350  
3         $330,600,000   $459,005,868  $1,403,013,963  
4         $317,000,000   $620,181,382  $1,316,721,747  
...                ...            ...             ...  
5777            $7,000             $0              $0  
5778            $6,000        $48,482        $240,495  
5779            $5,000         $1,338          $1,338  
5780            $1,400             $0              $0  
5781            $1,100       $181,041        $181,041  

[5782 rows x 5 columns]

Let's inspect the movie column and see where we have duplicates

In [24]:
print('The number of unique values in the column is:', movie_budgets_df['movie'].nunique(), 'out of a total of:', len(movie_budgets_df['movie']))

The number of unique values in the column is: 5698 out of a total of: 5782


In [25]:
# filter all the duplicated movie names and sort by the name to inspect the rows further
filt = movie_budgets_df.duplicated(subset=['movie'], keep=False)
movie_budgets_df.loc[filt].sort_values(by=['movie']).head(60)

release_date                         movie production_budget  \
4270  Dec 23, 1954  20,000 Leagues Under the Sea        $5,000,000   
5614  Dec 24, 1916  20,000 Leagues Under the Sea          $200,000   
1648  Apr 30, 2010     A Nightmare on Elm Street       $35,000,000   
5016   Nov 9, 1984     A Nightmare on Elm Street        $1,800,000   
2032  Nov 11, 1992                       Aladdin       $28,000,000   
80    May 24, 2019                       Aladdin      $182,000,000   
50     Mar 5, 2010           Alice in Wonderland      $200,000,000   
4759  Jul 28, 1951           Alice in Wonderland        $3,000,000   
4120  Oct 17, 1956   Around the World in 80 Days        $6,000,000   
340   Jun 16, 2004   Around the World in 80 Days      $110,000,000   
134   Mar 17, 2017          Beauty and the Beast      $160,000,000   
2485  Nov 13, 1991          Beauty and the Beast       $20,000,000   
3007  Nov 18, 1959                       Ben-Hur       $15,000,000   
422   Aug 19, 2016                       Ben-Hur       $95,000,000   
3307  Aug 14, 2015                      Brothers       $13,000,000   
2140   Dec 4, 2009                      Brothers       $26,000,000   
4058  Mar 26, 2010                          Ca$h        $7,000,000   
3231  Dec 31, 2008                          Ca$h       $13,700,000   
1925  Oct 18, 2013                        Carrie       $30,000,000   
5015  Nov 16, 1976                        Carrie        $1,800,000   
359   Nov 17, 2006                 Casino Royale      $102,000,000   
3381  Apr 28, 1967                 Casino Royale       $12,000,000   
3325   Apr 2, 1982                    Cat People       $12,500,000   
5668  Nov 16, 1942                    Cat People          $134,000   
4775  Feb 15, 1950                    Cinderella        $2,900,000   
412   Mar 13, 2015                    Cinderella       $95,000,000   
3050  Jun 12, 1981           Clash of the Titans       $15,000,000   
271    Apr 1, 2010           Clash of the Titans      $125,000,000   
475   Aug 19, 2011           Conan the Barbarian       $90,000,000   
2543  May 14, 1982           Conan the Barbarian       $20,000,000   
3978   May 6, 2005                         Crash        $7,303,082   
3669   Oct 4, 1996                         Crash       $10,000,000   
3366  Feb 15, 2002                    Crossroads       $12,000,000   
5485  Jun 23, 2015                    Crossroads          $500,000   
2330   Nov 9, 2012            Dangerous Liaisons       $24,200,000   
3187  Dec 21, 1988            Dangerous Liaisons       $14,000,000   
5074  Apr 20, 1979              Dawn of the Dead        $1,500,000   
2047  Mar 19, 2004              Dawn of the Dead       $28,000,000   
2851   Apr 8, 2008               Day of the Dead       $18,000,000   
4609  Jun 30, 1985               Day of the Dead        $3,500,000   
2463  Apr 16, 2010            Death at a Funeral       $21,000,000   
2652  Aug 17, 2007            Death at a Funeral       $20,000,000   
484    Jul 8, 2005                Fantastic Four       $87,500,000   
297    Aug 7, 2015                Fantastic Four      $120,000,000   
2610  Sep 29, 2017                    Flatliners       $20,000,000   
2127  Aug 10, 1990                    Flatliners       $26,000,000   
3860  Feb 17, 1984                     Footloose        $8,200,000   
2338  Oct 14, 2011                     Footloose       $24,000,000   
2873  Feb 13, 2009               Friday the 13th       $17,000,000   
5424   May 9, 1980               Friday the 13th          $550,000   
1835   Jun 8, 1984                  Ghostbusters       $30,000,000   
203   Jul 15, 2016                  Ghostbusters      $144,000,000   
273   May 19, 1998                      Godzilla      $125,000,000   
140   May 16, 2014                      Godzilla      $160,000,000   
3014  Aug 31, 2007                     Halloween       $15,000,000   
5536  Oct 17, 1978                     Halloween          $325,000   
3535  Oct 19, 20

In [26]:
#further inspection to see whether we have instances of having both the movie and release_date of one record duplicated 
filt2 = movie_budgets_df.duplicated(subset=['movie','release_date'], keep=False)
movie_budgets_df.loc[filt2].sort_values(by=['movie']).head(60)

Empty DataFrame
Columns: [release_date, movie, production_budget, domestic_gross, worldwide_gross]
Index: []

As seen above, even though the names have been duplicated, these are different movies based on the release date. We don't have instances where the name and release date are both similar in one record.  So it's safe to assume that the recent dated movies are remakes of the older ones hence we keep the records. 

The prodcution_budget, domestic_gross and worldwide_gross columns are not integers. This will make computation quite difficult.

In [27]:
#Inspection of the values shows that the values are strings
movie_budgets_df['domestic_gross'].apply(lambda x: type(x))

0       <class 'str'>
1       <class 'str'>
2       <class 'str'>
3       <class 'str'>
4       <class 'str'>
            ...      
5777    <class 'str'>
5778    <class 'str'>
5779    <class 'str'>
5780    <class 'str'>
5781    <class 'str'>
Name: domestic_gross, Length: 5782, dtype: object

We will have to get rid of the $ sign, the commas and convert the values to integers

In [28]:
# get rid of the $ sig, commas and converting the values into integers and the reassigning the columns
movie_budgets_df['domestic_gross'] = movie_budgets_df['domestic_gross'].apply(lambda x: x[1:]).str.replace(',', '').astype('int64')
movie_budgets_df['production_budget'] = movie_budgets_df['production_budget'].apply(lambda x: x[1:]).str.replace(',', '').astype('int64')
movie_budgets_df['worldwide_gross'] = movie_budgets_df['worldwide_gross'].apply(lambda x: x[1:]).str.replace(',', '').astype('int64')

In [29]:
#checking to see if the values have been changed to integers
movie_budgets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   release_date       5782 non-null   object
 1   movie              5782 non-null   object
 2   production_budget  5782 non-null   int64 
 3   domestic_gross     5782 non-null   int64 
 4   worldwide_gross    5782 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 226.0+ KB


In [30]:
#further inspecting the values to see where we have 0 as the values in the worldwide_gross and production_budget columns
filt = (movie_budgets_df['production_budget'] == 0) | (movie_budgets_df['worldwide_gross'] == 0)
missing_gross_amounts_df = movie_budgets_df.loc[filt]
print('The number of records with either the production_budget or world_wide gross as 0 are:', len(missing_gross_amounts_df))
missing_gross_amounts_df

The number of records with either the production_budget or world_wide gross as 0 are: 367


release_date                 movie  production_budget  domestic_gross  \
194   Dec 31, 2020              Moonfall          150000000               0   
479   Dec 13, 2017                Bright           90000000               0   
480   Dec 31, 2019      Army of the Dead           90000000               0   
535   Feb 21, 2020      Call of the Wild           82000000               0   
670   Aug 30, 2019             PLAYMOBIL           75000000               0   
...            ...                   ...                ...             ...   
5761  Dec 31, 2014  Stories of Our Lives              15000               0   
5764  Dec 31, 2007           Tin Can Man              12000               0   
5771  May 19, 2015      Family Motocross              10000               0   
5777  Dec 31, 2018                Red 11               7000               0   
5780  Sep 29, 2015  A Plague So Pleasant               1400               0   

      worldwide_gross  
194                 0  
479                 0  
480                 0  
535                 0  
670                 0  
...               ...  
5761                0  
5764                0  
5771                0  
5777                0  
5780                0  

[367 rows x 5 columns]

In [31]:
#Determining what percentage of rows in this columns contain missing values
total_num_rows = len(movie_budgets_df)
total_num_missing_rows = len(missing_gross_amounts_df)
percentage = total_num_missing_rows/total_num_rows
percentage

0.06347284676582497

6.3% of the values are zeros. These are not values that we can just fill since they are unique. 6.3% is not much so it's best
to just drop these rows.